node/network/dispute-distribution/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+
+	/// Authority discovery service.
+	authority_discovery: AD,
+
+	/// Metrics for this subsystem.
+	metrics: Metrics,
+}
+
+impl<Context, AD> Subsystem<Context> for DisputeDistributionSubsystem<AD>
+where
+	Context: SubsystemContext<Message = DisputeDistributionMessage> + Sync + Send,
+	AD: AuthorityDiscovery + Clone,
+{
+	fn start(self, ctx: Context) -> SpawnedSubsystem {
+		let future = self
+			.run(ctx)
+			.map_err(|e| SubsystemError::with_origin("dispute-distribution", e))
+			.boxed();
+
+		SpawnedSubsystem {
+			name: "dispute-distribution-subsystem",
+			future,
+		}
+	}
+}
+
+impl<AD> DisputeDistributionSubsystem<AD> 
+where
+	AD: AuthorityDiscovery + Clone,
+{
+	/// Create a new instance of the availability distribution.
+	pub fn new(keystore: SyncCryptoStorePtr, authority_discovery: AD, metrics: Metrics) -> Self {
+		let runtime = RuntimeInfo::new_with_config(runtime::Config {
+			keystore: Some(keystore),
+			session_cache_lru_size: DISPUTE_WINDOW as usize,
+		});
+		let (tx, sender_rx) = mpsc::channel(1);
+		let disputes_sender = DisputeSender::new(tx, metrics.clone());
+		Self { runtime, disputes_sender, sender_rx, authority_discovery, metrics }
+	}
+
+	/// Start processing work as passed on from the Overseer.
+	async fn run<Context>(mut self, mut ctx: Context) -> std::result::Result<(), Fatal>
+	where
+		Context: SubsystemContext<Message = DisputeDistributionMessage> + Sync + Send,
+	{
+		loop {
+			let message = MuxedMessage::receive(&mut ctx, &mut self.sender_rx).await;
+			match message {
+				MuxedMessage::Subsystem(result) => {
+					let result = match result? {
+						FromOverseer::Signal(signal) => {
+							match self.handle_signals(&mut ctx, signal).await {
+								Ok(SignalResult::Conclude) => return Ok(()),
+								Ok(SignalResult::Continue) => Ok(()),
+								Err(f) => Err(f),
+							}
+						}
+						FromOverseer::Communication { msg } =>
+							self.handle_subsystem_message(&mut ctx, msg).await,
+					};
+					log_error(result, "on FromOverseer")?;
+				}
+				MuxedMessage::Sender(result) => {
+					self.disputes_sender.on_task_message(
+						result.ok_or(Fatal::SenderExhausted)?
+					)
+					.await;
+				}
+			}
+		}
+	}
+
+	/// Handle overseer signals.
+	async fn handle_signals<Context: SubsystemContext> (
+		&mut self,
+		ctx: &mut Context,
+		signal: OverseerSignal
+	) -> Result<SignalResult>
+	{
+		match signal {
+			OverseerSignal::Conclude =>
+				return Ok(SignalResult::Conclude),
+			OverseerSignal::ActiveLeaves(update) => {
+				self.disputes_sender.update_leaves(
+					ctx,
+					&mut self.runtime,
+					update
+				)
+				.await?;
+			}
+			OverseerSignal::BlockFinalized(_,_) => {}
+		};
+		Ok(SignalResult::Continue)
+	}
+
+	/// Handle `DisputeDistributionMessage`s.
+	async fn handle_subsystem_message<Context: SubsystemContext> (
+		&mut self,
+		ctx: &mut Context,
+		msg: DisputeDistributionMessage
+	) -> Result<()>
+	{
+		match msg {
+			DisputeDistributionMessage::SendDispute(dispute_msg) =>
```

Comment:
The dispute coordinator doesn't appear to send this message. What does?

Reply:
If that's beyond the scope of this PR because it depends on the further changes you intend to make, please file a follow-up issue.

Reply:
Yes I have those ready and planned on doing a follow up, but it might actually be better to include those changes right away.